### PR TITLE
whisper: periodically rebuild bloom filter to remove stale topics

### PR DESCRIPTION
**Pull request by dshulyak**
In the current version of whisper, there is no way to remove a topic from
bloom filter after it was added. It leads to unnecessary accumulation of
such topics over time and undesirable network traffic.

In this change, I added an option to periodically rebuild bloom filter
from available filter topics as there is no way to simply revert an update
to bloom filter. But it would be possible if whisper used cuckoo filters.

Current interaction between peers allows resetting bloom filter to any
desirable state, allowing message from old bloom filter for certain time.
At the moment this time is 10s.

Closes: https://github.com/ethereum/go-ethereum/issues/16086

### CODE REVIEW

**Comment by gballet** [comment]
restarted the build to see if it passes

File: whisper/whisperv6/filter.go [hunk 1]
```diff
@@ -85,6 +85,15 @@ func (fs *Filters) Install(watcher *Filter) (string, error) {
 	return id, err
 }
 
+// Each is a thread safe iterator over all existing filters.
+func (fs *Filters) Each(f func(filter *Filter)) {
```
Comment (gballet) [inline whisper/whisperv6/filter.go]: Either change all instances of `range fs.watchers` or do not use this function. I recommend not using it, because it implies creating an anonymous functions, and we both know what that entails in terms of "Javascript"-ability of the code. We strive to keep the Whisper code as dumb as possible, to ensure it is easy to audit. In any case, this should be a refactoring and not part of this PR.

File: whisper/whisperv6/whisper.go [hunk 2]
```diff
@@ -868,13 +887,22 @@ func (whisper *Whisper) processQueue() {
...
+		defer bloomRebuild.Stop()
+		rebuildTicker = bloomRebuild.C
+	}
 	// Repeat updates until termination is requested
 	for {
 		select {
 		case <-expire.C:
 			whisper.expire()
-
+		case <-rebuildTicker:
```
Comment (gballet) [inline whisper/whisperv6/whisper.go]: The ticker is one way to do it, however it would be better to rebuild the filters on an as-needed basis. i.e. when the filters go above a given threshold or when they fall back under a given threshold. What do you think?

File: whisper/whisperv6/whisper.go [hunk 3]
```diff
@@ -553,6 +554,24 @@ func (whisper *Whisper) Subscribe(f *Filter) (string, error) {
...
 }
 
+// rebuildBloomFilter makes a new bloom filter from currently subscribed watchers
+// if new bloom filter is different from current bloom filter - peers will be notified
+func (whisper *Whisper) rebuildBloomFilter() {
+	if isFullNode(whisper.BloomFilter()) {
+		return
+	}
+	aggregate := make([]byte, bloomFilterSize)
+	whisper.filters.Each(func(filter *Filter) {
```
Comment (gballet) [inline whisper/whisperv6/whisper.go]: The thing is, as of https://github.com/ethereum/go-ethereum/pull/16176 we no longer remove filters automatically. I expect this to be way less effective. @gluk256 we should make filter removal an option, so that people who want to keep those filters can keep them, and those wanting a more unforgiving approach can do so as well.

**Comment by dshulyak** [comment]
@gballet thanks for review, though @gluk256 explained that it is designed like this for security reasons and clearing several bits in bloom filter should not be a huge bottleneck for light clients. I will get back to this PR if it proves to be an issue in the future

**Comment by gballet** [comment]
@dshulyak didn't know that, thanks for letting me know. Ping me if you decide to reopen it!
